### PR TITLE
feat(cli): `reth prune`

### PR DESCRIPTION
**Pull request by shekhirin**
Towards https://github.com/paradigmxyz/reth/issues/8995

This adds a CLI command `reth prune` that does two things:
1. Moves all data to static files
2. Prunes according to the configuration in `reth.toml`

The intended usage is to run this command mainly on a full node.

### CODE REVIEW

**Comment by shekhirin** [comment]
Tested it on the Holesky archive node by manually changing the `reth.toml` to make it full node -like. Here's a `diff` between two `db stats` runs.

```diff
12,13c12,13
< | AccountChangeSets          | 95554592  | 249872       | 1150462    | 0              | 5.3 GiB    |
< | AccountsHistory            | 32844661  | 11908        | 736471     | 7555           | 2.9 GiB    |
---
> | AccountChangeSets          | 593125    | 1106         | 9345       | 0              | 40.8 MiB   |
> | AccountsHistory            | 110757    | 244          | 7706       | 109            | 31.5 MiB   |
29c29
< | PruneCheckpoints           | 4         | 0            | 1          | 0              | 4 KiB      |
---
> | PruneCheckpoints           | 7         | 0            | 1          | 0              | 4 KiB      |
33,34c33,34
< | StorageChangeSets          | 143191275 | 301830       | 2216802    | 0              | 9.6 GiB    |
< | StoragesHistory            | 56757839  | 66107        | 2403603    | 4425           | 9.4 GiB    |
---
> | StorageChangeSets          | 892233    | 230          | 14171      | 0              | 56.3 MiB   |
> | StoragesHistory            | 520673    | 2268         | 41186      | 59             | 170 MiB    |
38c38
< | TransactionSenders         | 49871912  | 2083         | 466100     | 0              | 1.8 GiB    |
---
> | TransactionSenders         | 0         | 0            | 0          | 0              | 0 B        |
40c40
< | VersionHistory             | 35        | 1            | 2          | 0              | 12 KiB     |
---
> | VersionHistory             | 36        | 1            | 2          | 0              | 12 KiB     |
42,43c42,43
< | Tables                     |           |              |            |                | 62.1 GiB   |
< | Freelist                   | 511387    |              |            |                | 2 GiB      |
---
> | Tables                     |           |              |            |                | 33.4 GiB   |
> | Freelist                   | 8044745   |              |            |                | 30.7 GiB   |
```

File: bin/reth/src/commands/prune.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                receipts: stages_checkpoints[1],
+                transactions: stages_checkpoints[2],
+            })?;
+            static_file_producer.run(targets)?;
+            stages_checkpoints.into_iter().min().expect("exists")
+        };
+
+        // Deletes data which has been copied to static files.
+        if let Some(prune_tip) = lowest_static_file_height {
+            // Run the pruner according to the configuration, and don't enforce any limits on it
```
Comment (mattsse) [inline bin/reth/src/commands/prune.rs:48]: this should emit some info! log

File: bin/reth/src/commands/prune.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                    provider.get_stage_checkpoint(stage).map(|c| c.map(|c| c.block_number))
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+
+            let targets = static_file_producer.get_static_file_targets(HighestStaticFiles {
+                headers: stages_checkpoints[0],
+                receipts: stages_checkpoints[1],
+                transactions: stages_checkpoints[2],
+            })?;
+            static_file_producer.run(targets)?;
```
Comment (mattsse) [inline bin/reth/src/commands/prune.rs:42]: also info! log

**Review by mattsse — APPROVED** [review]
this looks simple enough,

defer to @joshieDo on details

File: bin/reth/src/commands/prune.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    pub async fn execute(self) -> eyre::Result<()> {
+        let Environment { config, provider_factory, .. } = self.env.init(AccessRights::RW)?;
+        let prune_config = config.prune.unwrap_or_default();
+
+        let static_file_producer =
+            StaticFileProducer::new(provider_factory.clone(), prune_config.segments.clone());
+        let static_file_producer = static_file_producer.lock();
+
+        // Copies data from database to static files
+        let lowest_static_file_height = {
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply bin/reth/src/commands/prune.rs]: done

File: crates/stages/api/src/pipeline/mod.rs [hunk 4]
```diff
@@ -248,26 +247,9 @@ where
...
-            let targets = static_file_producer.get_static_file_targets(HighestStaticFiles {
-                headers: stages_checkpoints[0],
-                receipts: stages_checkpoints[1],
-                transactions: stages_checkpoints[2],
-            })?;
-            static_file_producer.run(targets)?;
-            stages_checkpoints.into_iter().min().expect("exists")
-        };
+        let lowest_static_file_height =
+            self.static_file_producer.lock().copy_to_static_files()?.min();
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply crates/stages/api/src/pipeline/mod.rs:252]: yep, we don't use the producer beyond this point

**Review by joshieDo — APPROVED** [review]

**Pull request merged**